File: python/string_utils.py
import re


def is_valid_email(value):
    return bool(re.fullmatch(r"[^\s@]+@[^\s@]+\.[^\s@]+", value))


def is_numeric(value):
    try:
        float(value)
        return True
    except (TypeError, ValueError):
        return False


def normalize_whitespace(value):
    return " ".join(value.split())


def to_title_case(value):
    return value.strip().title()

File: javascript/string_utils.js
export const isValidEmail = (value) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value);

export const isNumeric = (value) => value !== '' && Number.isFinite(Number(value));

export const toTitleCase = (value) => value
  .trim()
  .toLowerCase()
  .replace(/\b\w/g, (character) => character.toUpperCase());

export const normalizeWhitespace = (value) => value.trim().replace(/\s+/g, ' ');
